ynput/ayon-shotgrid
# `settings` Add a map for entity's custom attributes

## Description
Add two new settings to the AYON-Shotgrid integration,:
* "Shotgrid Enabled Entities": User can define which entities they have enabled in their instance. 
* "Folder Attributes Map": User can define which AYON folder attributes map to Shotgrid entity fields. i.e.: `clipIn` -> `sg_cut_in`.

![image](https://github.com/ynput/ayon-shotgrid/assets/4348536/4b7bb6eb-8928-40fd-98b5-4bb0a754c713)
(image originally posted by @fabiaserra at #79 )

Also removed some unnecessary logging, and modify the services to make use of the new settings, bug fixes and quality of life improvements.

## Credits
Big shoutout to @fabiaserra and @Minkiu for their awesome work on this PR! Couldn't have done it without them!

## Changelog
3b3b424 `services` Remove SG `"Version"` sync We do not want to synchronise, and this is delegated to the publishing step, to create it in both AYON and SG.
5a2a8a9 `server.settings` Add `ShotgridCompatibility` A new section in the Shotgrid settings, where two new settings have been  added:  * "Shotgrid Enabled Entities": User can define which entities they have     enabled in their instance.  * "Folder Attributes Map": User can define which AYON folder attributes     map to Shotgrid entity fields. i.e.: `clipIn` -> `sg_cut_in`.
a0a4b98 `shotgrid_common.utils` Follow EntityHub's struc
7afd7fa `services` Adapt to new EntityHub struct Modified all places where we were accessing information that now lives in either `attribs` or `data` in the resulting `dict` in the new `shotgrid_common.utils._sg_to_ay_dict` which follows the EntityHub structure more closely.
09e7883 `services` Honour "Folder Attributes Map" setting
9e4e73e `services` Honour "Shotgrid Enabled Entities" setting
4e36120 `ayon_shotgrid_hub` Fix when matching hierarchies
854510d `ayon_shotgrid_hub` Reduce checks before sync Remove a check for missing attributes when initializing the Hub, which was adding noise and potential confusion to the user.
8904556 `services.processor` Pass `ShotgridProcessor` instance to handlers We were passing a lot of arguments to the `processor/handlers` this changes simplifies things a bit by just sending the actual class, so we then access the necessary attributes when creating the `AyonShotgridHub`.

## Todo
- [ ] Documentation, this needs to include tutorial on how to configure the addon and anatomy preset > this PR
- [x] Establishing minimal defaults for Attribute mapping. Those has to be created on SG if they don't exists > this PR
- [x] Investigate update and sync statuses (entity hub doesn't seem to allow this yet?) > next PR

> [!WARNING]  
> Due to new Settings Anatomy Presets enumerator your server has to be at least in **1.0.8+202403281636** version.

## Testing notes
> [!IMPORTANT] 
> First of all you need to bump version to any higher then it is in `version.py` and all services `pyproject.toml`. I am setting it up to **0.4.0** just to make it more visible but it is possible to set it to what ever higher then current version.

1.a Create package and install it to your server via Bundles. You can also use following terminal command to do it in you locally runed server docker container. This requires https://github.com/ynput/ayon-shotgrid/pull/78 to be merged into your local branch. But you can also just drop the created zip file to install addon section at bundles on server.
`python .\create_package.py --skip-zip --output [PATH TO REPOSITORY]\ayon-docker\addons` (requires #78)

2.1 Reboot your server so the version is available. 

2.2 Build Services. This requires to merge https://github.com/ynput/ayon-shotgrid/pull/74 into your testing local branch:
	1. `cd ./services` 
	2. `.\manage.ps1 build-all` - this will build all services as images to your docker 

2.3 Go to your Ayon services and Create Leecher, Processor, Transmitter services in the version 0.4.0 with appropriate bundle settings (I am using developer bundles and works fine to me)

3. You need to configure this version also in **production** bundle since frontend page is only available via production bundle. So it also means you have to configure settings in production, even you want to work with developer mode activated and do your tests in developer bundle.

4. You will need to go to Anatomy presets and create new one **SG** preset. Remove keep only one Folder and Task type. Ideally one which will be found in ShotGrid. So Shot folder type is ok and Comp(cmp) task type also good. 

5. Go to ShotGrid addon settings and set Anatomy preset to the **SG** (ayon+settings://shotgrid/anatomy_preset)

6. Attributes will be automatically aggregated from available folder attributes. It is important to define SG attributes without `sg_`  prefix so it is not doubled. Change type of attribute and scope in which you wish the SG attribute will be created if it does not exists. 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/a4848077-b4db-44ec-b9a6-e20c4fb821b2)

7. In Shotgrid tab at Studio Settings go to your project and sync Shotgrid to Ayon again so all attributes are created if they do not exists. 

8. Go to Shots and Add one. Add task template and confirm 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/ea66743f-2c82-406e-9d47-69d561350f11)

9. Notice that shot was created in Ayon 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/714cbdd7-2b11-4824-8533-2529a7364e27)

10. Make sure your Push Anatomy attribute on project is activated so the two way sync is active. Save Anatomy changes!
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/ca36ad2e-df04-4d82-b084-9350e37c192e)

11. Change Clip In Clip out attributes in Ayon and notice the change in ShotGrid 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/421bef10-2600-439b-9660-a14057dbe8fb)

12. Go back to ShotGrid and reload page to see those changes you attributes (takes about 10sec) 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/1eef5dbc-2e2a-4122-8d08-149a02cbf48a)

13. Now change those attributes in ShotGrid and wait for while to see that changes were reflected back in Ayon 
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/a59e15c6-0c8c-4494-bbb8-6d315fd442f2)
![image](https://github.com/ynput/ayon-shotgrid/assets/40640033/53240669-a7f3-46b3-96f4-d844388b509e)




[AY-1004](https://app.clickup.com/t/6658547/AY-1004)
